apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/PgNumeric.cs

```diff
@@ -0,0 +1,487 @@
+        /// trailing digits are permitted.
+        /// </para>
+        /// </remarks>
+        /// <param name="text">The text to parse. Must not be null.</param>
+        /// <returns>The parsed value.</returns>
+        /// <exception cref="FormatException">The value could not be parsed as a <see cref="PgNumeric"/>.</exception>
+        public static PgNumeric Parse(string text)
+        {
+            string message = TryParseImpl(text, out var value);
+            return message == null ? value : throw new FormatException(message);
+        }
+
+        /// <summary>
+        /// Attempts to parse a textual representation of a <see cref="PgNumeric"/> value.
+        /// </summary>
+        /// <remarks>
+        /// See <see cref="Parse"/> for format details. This method will return <c>true</c> if and only if
+        /// <see cref="Parse"/> would return without an exception.
+        /// </remarks>
+        /// <param name="text">The text to parse. Must not be null.</param>
+        /// <param name="value">The parsed value, or 0 on failure.</param>
+        /// <returns><c>true</c> if <paramref name="text"/> was parsed successfully; <c>false</c> otherwise.</returns>
+        public static bool TryParse(string text, out PgNumeric value)
+        {
+            string message = TryParseImpl(text, out value);
+            return message == null;
+        }
+
+        /// <summary>
+        /// Returns the error message for a FormatException if parsing fails, or null for success.
+        /// </summary>
+        private static string TryParseImpl(string text, out PgNumeric value)
+        {
+            GaxPreconditions.CheckNotNull(text, nameof(text));
+
+            if (string.Equals(text, NotANumber, StringComparison.InvariantCultureIgnoreCase))
+            {
+                value = new PgNumeric(default, 0, true);
+                return null;
+            }
+
+            if (!s_validation.IsMatch(text))
+            {
+                value = default;
+                return "Text representation must be digits, containing an optional decimal point, and an optional leading '-' sign.";
+            }
+            int pointIndex = text.IndexOf('.');
+
+            // The raw value which will end up in the result, on success.
+            BigInteger rawValue;
+            int scale = 0;
+            // If we have '.', then we need to extract the value and scale.
+            if (pointIndex != -1)
+            {
+                int decimalPlaces = text.Length - pointIndex - 1;
+                text = text.Remove(pointIndex, 1);
+                scale = decimalPlaces;
+                rawValue = BigInteger.Parse(text, CultureInfo.InvariantCulture);
+            }
+            else
+            {
+                // We don't have '.'
+                rawValue = BigInteger.Parse(text, CultureInfo.InvariantCulture);
+            }
+            if (!IsRawValueInRange(rawValue))
```

Comment:
Again, scale is relevant here as we are not scaling the integer to the max scale possible.